### DISCUSSION TITLE
What stages actually use the .sdc file?

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#4477] [General]
I want to reduce the rebuilding when the SDC file changes in https://github.com/The-OpenROAD-Project/megaboom

What stages actually use the .sdc file?

- floorplan.tcl reads in and writes out the .sdc file, but I don't think floorplan.tcl itself uses the information in the .sdc file. What about other floorplan stages?
- placement uses(I think), but does not modify the .tcl file
- cts uses and writes out a new .tcl file (propagated clocks)
- routing uses(I think), but does not modify the .sdc file
- final uses(I think) the .sdc file, but does not modify the .sdc file

**Comment by maliberty** [comment]
floorplan.tcl calls check_setup which does use SDC info.  I don't know that it modifies the SDC though.

In principle steps that modify the netlist could update the SDC (ie move/copy a constraint on a pin that is being modified).  In practice we don't do that much today but I don't think we should preclude it.

**Reply by oharboe** [reply]
I see.

As a more general strategy, it would be interesting to convert the inputs for a time consuming step to a minimal canonical form.

This could avoid recalculating things for e.g. macro placement?

What information really goes into macro placement, most of the time, I imagine the results could be reused if the initial conditions are not materially different...
